### PR TITLE
Wiring for Custom Datadir

### PR DESCRIPTION
**Pull request by D33r-Gee**
This pull request builds upon previous issues (#390, #392, and #397) by re-introducing the functionality for users to specify a custom data directory (`datadir`) during the onboarding process.

The majority of changes in `qml/bitcoin.cpp` are mainly splitting the onboarding and node creation so that the user can set their preferred `datatdir` it's inspired by the `qt/bitcoin.cpp` however it uses the anonymous namespace instead of using a class. The intention came from feedback on https://github.com/bitcoin-core/gui-qml/pull/390

In the QML pages, the `node_model` context property is no longer been initialized during onboarding. We're now using `options_model` instead. This means the code has been updated to use the `options_model` for settings that were previously set with the `node_model` context property.

Support for the Android version is being added to this PR.

### CODE REVIEW

**Review by johnny9 — CHANGES_REQUESTED** [review]
https://github.com/bitcoin-core/gui-qml/actions/runs/9523738145/job/26255610402?pr=408
```
qml/models/options_model.cpp:232:9: error: use of undeclared identifier 'settings'
        settings.setValue("strDataDir", path);
        ^
```        

Android builds are failing. Will need that fixed.

My testing on desktop seemed to work well. I had one odd thing happening. After onboarding, I tried again with using -resetguisettings and the onboarding page had a strange Custom path as the initial selection. I imagine this is maybe just resetguisettings not clearing out the datadir in QSettings but haven't dug in yet.

Screenshot from 2024-06-14 23-26-36

**Comment by D33r-Gee** [comment]
Thanks @johnny9 for testing
> Android builds are failing. Will need that fixed.

addressed it for now with the latest push [679818a](https://github.com/bitcoin-core/gui-qml/commit/679818a1d289f7d76f463b6deb13dd1c828eb486)
> My testing on desktop seemed to work well. I had one odd thing happening. After onboarding, I tried again with using -> > resetguisettings and the onboarding page had a strange Custom path as the initial selection. I imagine this is maybe just resetguisettings not clearing out the datadir in QSettings but haven't dug in yet.

Addressed the above for now... although there are a few issues when using the -resetguisettings flag, so a question I have is is it crucial to address them here or can they be addressed in later PRs that specifically tackle them?

File: src/qml/components/StorageLocations.qml [hunk 1]
```diff
@@ -22,11 +22,21 @@ ColumnLayout {
         ButtonGroup.group: group
         text: qsTr("Default")
         description: qsTr("Your application directory.")
-        customDir: optionsModel.getDefaultDataDirString
-        checked: optionsModel.dataDir === optionsModel.getDefaultDataDirString
+        checked: true
```
Comment (MarnixCroes) [inline src/qml/components/StorageLocations.qml:25]: `qrc:/qml/components/StorageLocations.qml:19:5: QML OptionButton: Binding loop detected for property "customDir"`

select custom datadir -> select default -> select custom datadir -> default keeps being selected in GUI

File: src/qml/bitcoin.cpp [hunk 2]
```diff
@@ -158,6 +159,178 @@ void setupChainQSettings(QGuiApplication* app, QString chain)
...
+    return true;
+    } else {
+        return false;
+    }
+}
+
+QGuiApplication* m_app;
+QQmlApplicationEngine* m_engine;
+// ArgsManager* m_gArgs;
+QQuickWindow* m_window;
```
Comment (MarnixCroes) [inline src/qml/bitcoin.cpp:178]: some are defined but no used
and give compile warning

```
qml/bitcoin.cpp:195:10: warning: ‘{anonymous}::m_introThread’ defined but not used [-Wunused-variable]
  195 | QThread* m_introThread;
      |          ^~~~~~~~~~~~~
qml/bitcoin.cpp:191:5: warning: ‘{anonymous}::m_prune_size_gb’ defined but not used [-Wunused-variable]
  191 | int m_prune_size_gb;
      |     ^~~~~~~~~~~~~~~
qml/bitcoin.cpp:190:6: warning: ‘{anonymous}::m_prune’ defined but not used [-Wunused-variable]
  190 | bool m_prune;
      |      ^~~~~~~
qml/bitcoin.cpp:178:15: warning: ‘{anonymous}::m_window’ defined but not used [-Wunused-variable]
  178 | QQuickWindow* m_window;
      |               ^~~~~~~~
```

**Review by MarnixCroes — COMMENTED** [review]
679818a1d289f7d76f463b6deb13dd1c828eb486

- during onboarding, when clicking `Custom datadir` it gives this output in terminal
```
Model size of -106 is less than 0
Model size of -6 is less than 0
```

- startNode and startOnboarding have a lot of duplicate code. maybe consider extracting common functionality to avoid duplication?

**Comment by D33r-Gee** [comment]
> [679818a](https://github.com/bitcoin-core/gui-qml/commit/679818a1d289f7d76f463b6deb13dd1c828eb486)

Thanks testing and reviewing @MarnixCroes 
> 
> * during onboarding, when clicking `Custom datadir` it gives this output in terminal
> 
> ```
> Model size of -106 is less than 0
> Model size of -6 is less than 0
> ```

Hmmm... not sure what that is... will look into it

> 
> * startNode and startOnboarding have a lot of duplicate code. maybe consider extracting common functionality to avoid duplication?

Yes some refinement can be done there... will look into how best approach it

**Comment by D33r-Gee** [comment]
Currently tackling a strange beahvior involving the custom `datadir` and reading the settings files:

when going through first setup it up it works great and the `settings.json` file is created in the the proper place... 

However on restart it seems like it doesn't read the `settings.json` file from the new `datadir` and only looks in the default directory. 

Right now trying to see how that can be addressed... will update soon with what I find

**Comment by D33r-Gee** [comment]
with [aebfc89](https://github.com/bitcoin-core/gui-qml/commit/aebfc89b7e983748e95708dbd7e088fba171b9ec) addressed the `settings.json` file not being read on restart by moving the `gArgs.ReadSettingsFile` right before node creation in `qml/bitcoin.cpp`

Now will tackle the `-resetguisettings` not updating the `datadir` location right away

**Comment by D33r-Gee** [comment]
This update [6d4c111](https://github.com/bitcoin-core/gui-qml/commit/6d4c111c4028525a4071cf03f6d73c8b2e579b8f) is the first pass at optimizing the qml/bitcoin.cpp code.

There are no major functionality updates, mainly code consolidation.

File: src/qml/components/StorageLocations.qml (hunk 1 above)
Comment (D33r-Gee) [reply src/qml/components/StorageLocations.qml:25]: done! Used new function to mitigate "default reset" issues

File: src/qml/bitcoin.cpp (hunk 2 above)
Comment (D33r-Gee) [reply src/qml/bitcoin.cpp]: Thanks for catching those!

removed unused member variables...

**Comment by D33r-Gee** [comment]
> * during onboarding, when clicking `Custom datadir` it gives this output in terminal
> 
> ```
> Model size of -106 is less than 0
> Model size of -6 is less than 0
> ```

@MarnixCroes are you still seeing the above pop-up with [6d4c111](https://github.com/bitcoin-core/gui-qml/commit/6d4c111c4028525a4071cf03f6d73c8b2e579b8f)?

**Review by pablomartin4btc — COMMENTED** [review]
tACK 6d4c111c4028525a4071cf03f6d73c8b2e579b8f

Normal use case, running `bitcoin-qt` and setting the custom data dir in one go, works as expected.

A use case that has an issue I think, which is selecting a custom datadir, going back to the default, I can't manage to set a custom datadir back, the default remains selected and this is working in `main`.

I got this crash (not always but quite common) when I click on the "Custom" component:

```
./src/qt/bitcoin-qt -signet
Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.
**
Gtk:ERROR:../../../../gtk/gtkpathbar.c:1342:set_button_image_get_info_cb: assertion failed: (cancellable == data->button_data->cancellable)
Bail out! Gtk:ERROR:../../../../gtk/gtkpathbar.c:1342:set_button_image_get_info_cb: assertion failed: (cancellable == data->button_data->cancellable)
Aborted (core dumped)
```

To reproduce it delete any `settings.json` and any "`BitcoinCore-App*.conf`" in `home/.config/BitcoinCore`.

Code-wise I need more time to check the changes, especially the last commit, cannot be split? It would be useful to add a brief description of the intention, what's the intention of the changes in the protocol of `bitcoin.cpp` (re-structuring it for clarity? something was missing/ needed?).

On the first commit, the commit name doesn't explain much either (eg , it seems there's a refactoring within, what's the motivation behind making node a pointer instead of a reference? If you check the code in core/ qt it's always defined as &. Is there an issue you are trying to solve or just syntax or semantic?).

File: src/common/settings.h [hunk 3]
```diff
@@ -10,6 +10,7 @@
 #include <cstddef>
 #include <map>
 #include <string>
+#include <univalue.h>
```
Comment (pablomartin4btc) [inline src/common/settings.h:13]: Every commit should pass compilation and all tests so if the failure occurs on the previous commit you should include it there (we are missing the "`CI / test each commit (pull_request)`" that exists on the other repos and I think @hebasto has it pending with other things he wanted to merge here into this repo).

**Comment by MarnixCroes** [comment]
> > * during onboarding, when clicking `Custom datadir` it gives this output in terminal
> > 
> > ```
> > Model size of -106 is less than 0
> > Model size of -6 is less than 0
> > ```
> 
> @MarnixCroes are you still seeing the above pop-up with [6d4c111](https://github.com/bitcoin-core/gui-qml/commit/6d4c111c4028525a4071cf03f6d73c8b2e579b8f)?

I do not

File: src/qml/components/StorageLocations.qml (hunk 1 above)
Comment (MarnixCroes) [reply src/qml/components/StorageLocations.qml:25]: > default keeps being selected in GUI

this is still an issue 
(6d4c111c4028525a4071cf03f6d73c8b2e579b8f)

**Comment by D33r-Gee** [comment]
Thanks @pablomartin4btc for testing and for your questions.

> what's the intention of the changes in the protocol of `bitcoin.cpp` (re-structuring it for clarity?

The majority of changes in qml/bitcoin.cpp are mainly splitting the onboarding and node creation so that the user can set their preferred `datatdir` it's inspired by the qt/bitcoin.cpp however it uses the anonymous namespace instead of using a class. The intention came from feedback on #390 

>  what's the motivation behind making node a pointer instead of a reference?

I'm assuming you are referencing: https://github.com/bitcoin-core/gui-qml/blob/f3b8c5a625a64ab490f9ba271c5310f988c5a0ae/src/qml/models/options_model.cpp#L27

if so the logic there came from having to initiate the `options_model` in `qml/bitcoin.cpp` with the node as a `nullptr`. And this was a way for it to work with the least amount of changes.

> If you check the code in core/ qt it's always defined as &. Is there an issue you are trying to solve or just syntax or semantic?).

Yes in `qt/bitcoin.cpp` the OptionsModel is initiated through a reference because node creation doesn't happen until the splashscreen/intro have gone through and no settings are being set with the OptionsModel.

On the other hand in `qml/bitcoin.cpp` the `options_model` is needed for onboarding so initiating it with a `nullptr` is allowing the QML context property to be set without `Node` initiation.

File: src/qml/models/nodemodel.h [hunk 4]
```diff
@@ -7,7 +7,7 @@
 
 #include <interfaces/handler.h>
 #include <interfaces/node.h>
-#include <clientversion.h>
+// #include <clientversion.h>
```
[the remark this answers is not in the document]
Comment (D33r-Gee) [reply src/qml/models/nodemodel.h:10]: left in case there's another way... removing it on the next push

File: src/qml/components/StorageLocations.qml (hunk 1 above)
Comment (D33r-Gee) [reply src/qml/components/StorageLocations.qml:25]: Thanks for checking this... The issue is now addressed in [7a34b0d](https://github.com/bitcoin-core/gui-qml/commit/7a34b0da39c0fe63b56981787a8177f492febd1a)

File: src/qml/models/nodemodel.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (D33r-Gee) [reply src/qml/models/nodemodel.h]: commented out sections are now deleted with [7a34b0d](https://github.com/bitcoin-core/gui-qml/commit/7a34b0da39c0fe63b56981787a8177f492febd1a)

**Review by MarnixCroes — COMMENTED** [review]
7a34b0da39c0fe63b56981787a8177f492febd1a
did some manual testing, so far works ok

some commits could be squashed imo, also as Pablo mentioned here https://github.com/bitcoin-core/gui-qml/pull/408#discussion_r1658350670

**Comment by D33r-Gee** [comment]
Thanks @MarnixCroes for testing and @hebasto for the comments.

Will update once more with cleaner and clearer commits... 

Also working on the Android version... Found a workaround for QSettings not working with Android, will post Android related commits soon...

File: src/qml/components/AboutOptions.qml [hunk 5]
```diff
@@ -51,7 +51,8 @@ ColumnLayout {
         header: qsTr("Version")
         actionItem: ExternalLink {
             parentState: versionLink.state
-            description: nodeModel.fullClientVersion
+            // description: nodeModel.fullClientVersion
```
[the remark this answers is not in the document]
Comment (D33r-Gee) [reply src/qml/components/AboutOptions.qml]: removed with latest push

File: src/qml/models/options_model.cpp [hunk 6]
```diff
@@ -168,12 +242,29 @@ bool OptionsQmlModel::setCustomDataDirArgs(QString path)
     QString newPrefix = "/storage/self/primary/";
     QString path = uri.replace(originalPrefix, newPrefix);
 #else
+    QSettings settings;
     path = QUrl(path).toLocalFile();
 #endif // __ANDROID__
-        qDebug() << "PlaceHolder: Created data directory: " << path;
+        // Delet before merge
```
[the remark this answers is not in the document]
Comment (D33r-Gee) [reply src/qml/models/options_model.cpp]: removed it with latest update

**Comment by D33r-Gee** [comment]
This update [2cf64dd](https://github.com/bitcoin-core/gui-qml/commit/2cf64dd91420b6b18af3af14f249956b2e4167ee) introduces Android functionality by creating qml/androidcustomdatadir files. This allows the app to retrieve the custom data directory path, if previously set by the user.

File: src/qml/bitcoin.cpp [hunk 7]
```diff
@@ -158,6 +160,188 @@ void setupChainQSettings(QGuiApplication* app, QString chain)
...
+    if(fs::exists(GUIUtil::QStringToPath(strDataDir))){
+        gArgs.SoftSetArg("-datadir", fs::PathToString(GUIUtil::QStringToPath(strDataDir)));
+        gArgs.ClearPathCache();
+    return true;
+    } else {
+        return false;
+    }
+}
+
+QGuiApplication* m_app;
```
Comment (johnny9) [inline src/qml/bitcoin.cpp:176]: statically defining all of these pointers is concerning. Eventhough it is redundant it feels safer to keep a lot of these withing the function scopes.

Comment (D33r-Gee) [reply src/qml/bitcoin.cpp:176]: Ok thanks for looking into it... will refactor further to use arguments instead of pointers

Comment (D33r-Gee) [reply src/qml/bitcoin.cpp:176]: After our online chat, let's keep things as is with the caveat that if any issues arise due to the static pointers we'll have to refactor further. Resolving for now...

**Review by johnny9 — APPROVED** [review]
tACK 2cf64dd91420b6b18af3af14f249956b2e4167ee

Onboarded with a custom directory, started IBD, then restarted the application to make sure the application continued to use the custom directory. Worked on both Linux Desktop and Android.

**Comment by hebasto** [comment]
The https://github.com/bitcoin-core/gui-qml/pull/408/files#r1658350670 still needs to be addressed.

File: src/qml/bitcoin.cpp [hunk 8]
```diff
@@ -93,7 +95,7 @@ AppMode SetupAppMode()
         wallet_enabled = false;
     #endif // ENABLE_WALLET
 
-    return AppMode(mode, wallet_enabled);
+    return new AppMode(mode, wallet_enabled);
```
Comment (hebasto) [inline src/qml/bitcoin.cpp:98]: Why this change is required? Is it safe?

File: src/qml/bitcoin.cpp (hunk 7 above)
Comment (hebasto) [reply src/qml/bitcoin.cpp:176]: > After our online chat...

Could the main points from that discussion be posted here?

File: src/qml/bitcoin.cpp [hunk 9]
```diff
@@ -158,6 +160,188 @@ void setupChainQSettings(QGuiApplication* app, QString chain)
...
+    m_engine->rootContext()->setContextProperty("optionsModel", m_options_model);
+
+    m_node_model->startShutdownPolling();
+
+    return true;
+}
+
+void startNodeAndTransitionSlot() { createNode(*m_app, *m_engine, m_argc, m_argv, gArgs); }
+
+// moved this function to a separate function for code reusability
```
Comment (hebasto) [inline src/qml/bitcoin.cpp:259]: Comments like that seem add no value. However, this sentence is more suitable for a commit message.

File: src/qml/bitcoin.cpp [hunk 10]
```diff
@@ -158,6 +160,188 @@ void setupChainQSettings(QGuiApplication* app, QString chain)
         app->setApplicationName(QAPP_APP_NAME_REGTEST);
     }
 }
+
+// added this function to set custom data directory when starting node
```
Comment (hebasto) [inline src/qml/bitcoin.cpp:164]: Comments like that seem add no value. However, this sentence is more suitable for a commit message.

File: src/qml/bitcoin.cpp (hunk 7 above)
Comment (D33r-Gee) [reply src/qml/bitcoin.cpp:176]: To summarize our convo: We talked about the safety of static pointers in the code. We thought about adding a class like in #390, however decided to leave them for now, since there might be [issues](https://github.com/bitcoin-core/gui-qml/pull/390#discussion_r1531658261) raised by @Jarold that we need to consider first. That said we both agreed that having a clear owner for pointers is usually a good idea for better code management. Is there anything I missed @johnny9 ?

File: src/qml/bitcoin.cpp (hunk 8 above)
Comment (D33r-Gee) [reply src/qml/bitcoin.cpp:98]: Looking into it...

**Comment by D33r-Gee** [comment]
With the latest push [28ab2a0](https://github.com/bitcoin-core/gui-qml/commit/28ab2a0a06571f4ec500fa500ec0cab738dd0694) reverted the `AppMode` definition in `qml/bitcoin.cpp` to the original. Also cleaned out comments based on @hebasto feedback.

File: src/qml/bitcoin.cpp (hunk 8 above)
Comment (D33r-Gee) [reply src/qml/bitcoin.cpp]: Thanks for catching that, no it is not safe! While refactoring ran into some compiling errors and thought (mistakenly) that turning `AppMode` into a pointer would be the way to go. That’s not necessary so reverting back to the original definition.

File: src/qml/bitcoin.cpp (hunk 9 above)
Comment (D33r-Gee) [reply src/qml/bitcoin.cpp]: Removed. Thanks for pointing that out.

File: src/qml/bitcoin.cpp (hunk 10 above)
Comment (D33r-Gee) [reply src/qml/bitcoin.cpp]: Also removed. Thanks for pointing it out.

**Review by MarnixCroes — COMMENTED** [review]
28ab2a0a06571f4ec500fa500ec0cab738dd0694

- the used `Data Directory` indicated in the GUI can be different from the actual used data direcory: run with resetguisettings -> select custom data dir -> let it create some of the files (don't have to finish IBD) -> shutdown -> delete the data (not the dir) -> launch app again (without resetguisettings) -> it uses the previous data dir (as supposed to) but in the Storage Settings it indicates it is using the default data dir which is inccorect as it is not using this.
- I've had twice this crash, which Pablo also encountered earlier
I did not delete any files, just ran and clicked on the `Custom`
```
Gtk:ERROR:../../../../gtk/gtkpathbar.c:1342:set_button_image_get_info_cb: assertion failed: (cancellable == data->button_data->cancellable)
Bail out! Gtk:ERROR:../../../../gtk/gtkpathbar.c:1342:set_button_image_get_info_cb: assertion failed: (cancellable == data->button_data->cancellable)
Aborted (core dumped)
```

**Comment by D33r-Gee** [comment]
Thanks @MarnixCroes for testing and discovering:
> * the used `Data Directory` indicated in the GUI can be different from the actual used data direcory: run with resetguisettings -> select custom data dir -> let it create some of the files (don't have to finish IBD) -> shutdown -> delete the data (not the dir) -> launch app again (without resetguisettings) -> it uses the previous data dir (as supposed to) but in the Storage Settings it indicates it is using the default data dir which is inccorect as it is not using this.

I've addressed it for both desktop (tested on Ubuntu 22.04) and Android with the latest push [834d8ac](https://github.com/bitcoin-core/gui-qml/commit/834d8ac2e03d8acd55f2c686fc9e1f4b5e236659)

Now the address persist even in the case of data deletion and not not the directory.

**Comment by D33r-Gee** [comment]
> I've had twice this crash, which Pablo also encountered earlier
I did not delete any files, just ran and clicked on the Custom

haven't been able to repro will investigate further...

**Comment by D33r-Gee** [comment]
with the latest push [b67ae97](https://github.com/bitcoin-core/gui-qml/commit/b67ae97f5091726db2694d591ffb342db8ebcd1a) got rid off the check mark happening when clicking the `custom` button in  `StorageLocations.qml`

@MarnixCroes please let me know if that fixes the `GTK` crash issue?

**Comment by MarnixCroes** [comment]
> with the latest push [b67ae97](https://github.com/bitcoin-core/gui-qml/commit/b67ae97f5091726db2694d591ffb342db8ebcd1a) got rid off the check mark happening when clicking the `custom` button in `StorageLocations.qml`
> 
> @MarnixCroes please let me know if that fixes the `GTK` crash issue?

I can still repro it on b67ae97f5091726db2694d591ffb342db8ebcd1a

File: src/qml/androidcustomdatadir.cpp [hunk 11]
```diff
@@ -0,0 +1,48 @@
+// Copyright (c) 2023 The Bitcoin Core developers
```
Comment (MarnixCroes) [inline src/qml/androidcustomdatadir.cpp:1]: nit 
```suggestion
// Copyright (c) 2024 The Bitcoin Core developers
```
or maybe even `2023-present`, which seems to be used in the main repo now.

**Review by MarnixCroes — COMMENTED** [review]
b67ae97f5091726db2694d591ffb342db8ebcd1a
the Data Directory in the GUI is also incorrect (displays the default one) when user runs with -datadir start up param

**Comment by D33r-Gee** [comment]
> [b67ae97](https://github.com/bitcoin-core/gui-qml/commit/b67ae97f5091726db2694d591ffb342db8ebcd1a) the Data Directory in the GUI is also incorrect (displays the default one) when user runs with -datadir start up param

Ah that's a great find... hadn't tested that... ok will update

File: src/qml/androidcustomdatadir.cpp (hunk 11 above)
Comment (D33r-Gee) [reply src/qml/androidcustomdatadir.cpp:1]: good catch...

**Comment by D33r-Gee** [comment]
updated to [94215e5](https://github.com/bitcoin-core/gui-qml/commit/94215e5427738c56a714d0c46a63b742e542f129) it now displays the proper `datadir` when using the `-datadir` arg

> I can still repro it on [b67ae97](https://github.com/bitcoin-core/gui-qml/commit/b67ae97f5091726db2694d591ffb342db8ebcd1a)

ok that's a bummer... will keep investigating although it's difficult to know since I'm not able to repro... will ask around

**Comment by D33r-Gee** [comment]
@MarnixCroes RE: the GTK crash, are you using the depends (static build) or are you using dynamic to compile?

**Comment by MarnixCroes** [comment]
> @MarnixCroes RE: the GTK crash, are you using the depends (static build) or are you using dynamic to compile?

yes using the depends
(Pablo also encountered it earlier https://github.com/bitcoin-core/gui-qml/pull/408#pullrequestreview-2146000827)

**Comment by D33r-Gee** [comment]
> > @MarnixCroes RE: the GTK crash, are you using the depends (static build) or are you using dynamic to compile?
> 
> yes using the depends (Pablo also encountered it earlier [#408 (review)](https://github.com/bitcoin-core/gui-qml/pull/408#pullrequestreview-2146000827))

ok when you run `dpkg -l "*gtk*" | grep ii` do you get this : `ii  qt5-gtk-platformtheme:amd64  5.15.3+dfsg-2ubuntu0.2       amd64        Qt 5 GTK+ 3 platform theme` ?

**Comment by MarnixCroes** [comment]
> ok when you run `dpkg -l "*gtk*" | grep ii` do you get this : `ii qt5-gtk-platformtheme:amd64 5.15.3+dfsg-2ubuntu0.2 amd64 Qt 5 GTK+ 3 platform theme` ?

yes

**Comment by hebasto** [comment]
Please rebase to resolve conflicts.

**Comment by D33r-Gee** [comment]
rebased over main with [8d53c44](https://github.com/bitcoin-core/gui-qml/commit/8d53c44ad99c7844762dde5caf3aba03339bb8f4)

tested out on WSL Ubuntu 22.04 however didn't tested out on Android...

File: src/qml/bitcoin.cpp [hunk 12]
```diff
@@ -160,6 +162,194 @@ void setupChainQSettings(QGuiApplication* app, QString chain)
...
+    QScopedPointer<const NetworkStyle> network_style{NetworkStyle::instantiate(Params().GetChainType())};
+    assert(!network_style.isNull());
+    m_engine->addImageProvider(QStringLiteral("images"), new ImageProvider{network_style.data()});
+
+    m_isOnboarded = true;
+
+    m_options_model = new OptionsQmlModel{nullptr, m_isOnboarded};
+    m_engine->rootContext()->setContextProperty("optionsModel", m_options_model);
+
+    // moved this so that the settings.json file is read and parsed before creating the node
```
Comment (MarnixCroes) [inline src/qml/bitcoin.cpp:314]: comment could be more general, without the "moved this"
same as https://github.com/bitcoin-core/gui-qml/pull/408#discussion_r1675628865

**Review by MarnixCroes — COMMENTED** [review]
8d53c44ad99c7844762dde5caf3aba03339bb8f4

apart from the know [crash issue](8d53c44ad99c7844762dde5caf3aba03339bb8f4) (which happens occasionally) the previous issues are fixed and it works as expected.

commit history could still be improved (commits like https://github.com/bitcoin-core/gui-qml/pull/408/commits/c1067d10740fd4c3fbba89445bcfd129893e3c0b) which was also previously mentioned https://github.com/bitcoin-core/gui-qml/pull/408/commits/c1067d10740fd4c3fbba89445bcfd129893e3c0b#r1658350670

File: src/qml/bitcoin.cpp (hunk 12 above)
Comment (D33r-Gee) [reply src/qml/bitcoin.cpp:314]: makes sense

**Comment by johnny9** [comment]
ACK 370bc643be09c49a945eed3ca13e792cfafb385e

tested the flow on Desktop and Android and both worked as expected without any crashing.

Desktop
Screenshot from 2024-09-17 15-23-46
Screenshot from 2024-09-17 15-23-30

Android:
Screenshot from 2024-09-17 00-11-56
Screenshot from 2024-09-17 00-12-12

**Review by pablomartin4btc — COMMENTED** [review]
tested d76761d30650e49751d055575fb4fbf996a27bb1 on Ubuntu 22.04.

I found a few inconsistencies regarding the commits structure:
- As I [noticed](https://github.com/bitcoin-core/gui-qml/pull/408#discussion_r1658350670) earlier, currently 4rd and 5th commits should be squashed together.
- Same for 2nd and 3rd commits, the 2nd commit fails as you removed `fullClientVersion` from the node model but you didn't update the reference until the next commit. 
- I'd put all the android changes together at the end and separate from non-android stuff (in fact, going forward, if possible, I'd add them as a follow-up PR to make the review and testing easier and faster), I don't see the reason of splitting them in so many, this also would avoid other mistakes: in commit *refactoring to allow..." you are including `src/qml/androidcustomdatadir.cpp` that is being added 2 commits later. Commit "android: permissions for custom datadir" is not for custom datadir specific but for android "external" storage in general.

I think the PR should have 4 commits in total (maybe 5 if you want to have the android storage permission one separate).

Regarding the UI and the wiring/ persistence of the datadir configuration, it works as expected. As I mentioned in the UI only PR earlier, we need to think about the different chain-nets or when user specify -datadir, perhaps this is more related with the onboarding, do we need to trigger the onboarding if it's the first time on those situations?

Somehow I got different numbers in the storage estimation:

image

image

As it can be seen as well, regarding the Gtk error, this time didn't crash but it appears after the 6th time of `bitcoin-qt` execution. 

`(bitcoin-qt:276616): Gtk-CRITICAL **: 23:44:35.670: gtk_image_set_from_gicon: assertion 'GTK_IS_IMAGE (image)' failed
`

This is happening when the file dialog pops up, started to happen in the UI only PR #397.

Please try to re-org the commits, let me know if you need any help on this, so I can do another code review ASAP and we can merge this as I also need to touch the options model, onboarding and `bitcoin.cpp`.

**Comment by D33r-Gee** [comment]
Thanks @johnny9, @MarnixCroes, @pablomartin4btc  and @hebasto  for testing and reviewing this PR.

@pablomartin4btc , thanks for pointing the commit inconsistencies I will keep your feedback in mind as I publish future PRs.

After speaking with @jarolrod  it is now clear that the code is not up to par (to put it politely).

I am closing it, so I can focus on AssumeUTXO integration, however I opened issue #426 so whoever takes it on can learn from my mistakes.
